signer/src/network/libp2p/bootstrap.rs

```diff
@@ -0,0 +1,446 @@
+                let (connection_id, error, peer_id) = (e.connection_id, e.error, e.peer_id);
+                if self.pending_connections.remove(&connection_id) {
+                    tracing::debug!(
+                        %connection_id,
+                        ?peer_id,
+                        %error,
+                        "failed to dial seed peer"
+                    );
+                }
+            }
+            _ => {}
+        }
+    }
+
+    fn on_connection_handler_event(
+        &mut self,
+        _peer_id: PeerId,
+        _connection_id: libp2p::swarm::ConnectionId,
+        _event: libp2p::swarm::THandlerOutEvent<Self>,
+    ) {
+    }
+
+    #[tracing::instrument(skip_all)]
+    fn poll(
+        &mut self,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<libp2p::swarm::ToSwarm<Self::ToSwarm, libp2p::swarm::THandlerInEvent<Self>>>
+    {
+        // If we have any pending events, we return them immediately.
+        if let Some(event) = self.pending_events.pop_front() {
+            return Poll::Ready(event);
+        }
+
+        // If we have any connections, we consider ourselves bootstrapped.
+        // Determine if there was a state change or not and return the correct
+        // poll result.
+        if self.connected_peer_count() >= 1 {
+            return match self.is_bootstrapped {
+                true => self.next_pending_event(),
+                false => {
+                    // We've just bootstrapped, so we flag our state as
+                    // bootstrapped and emit an event to the swarm.
+                    self.is_bootstrapped = true;
+                    let connection_count = self.connection_count();
+                    let connected_peers = self.connected_peers().iter().map(|p| p.0).collect();
+
+                    tracing::info!(%connection_count, ?connected_peers, "network bootstrapping complete");
+                    Poll::Ready(ToSwarm::GenerateEvent(BootstrapEvent::Bootstrapped {
+                        connection_count,
+                        connected_peers,
+                    }))
+                }
+            };
+        }
+
+        // If we're here then we're not connected to any peers. If our current
+        // state is bootstrapped, we need to re-bootstrap.
+        if self.is_bootstrapped {
+            tracing::info!(
+                "state is bootstrapped but not connected to any peers; re-bootstrapping is needed"
+            );
+            self.is_bootstrapped = false;
+            self.pending_events
+                .push_back(ToSwarm::GenerateEvent(BootstrapEvent::NoConnectedPeers));
+        }
+
+        // If we've attempted to bootstrap recently, we wait until the interval has passed.
+        if let Some(last_bootstrap) = self.last_attempted_at {
+            if last_bootstrap.elapsed() < self.config.bootstrap_interval {
+                return self.next_pending_event();
+            }
+        }
```

Comment:
Sanity check: we are going to get poll'd even if there's no swarm activity, right? I'm thinking if it could happen that we need to bootstrap again, but we consume all the events so if we don't get poll'd anymore we would be stuck.

Reply:
Yeah, _we_ are responsible for driving/polling the swarm, which we do at an interval in the event loop, so every time we do that will trigger a poll on all enabled behaviors.